### DISCUSSION TITLE
Voorstel om informatieobjecten te kunnen relateren

### DISCUSSION THREAD

**Discussion by HenriKorver** [VNG-Realisatie/gemma-zaken#2240] [Ideas]
# Voorstel om informatieobjecten onderling te kunnen relateren

Dit is een voorstel om issues #2078 en #2079 te tackelen.

De volgende referentietabel wordt opgenomen in de aanvullende specificatie van de Documenten API. In deze eerste versie bevat de tabel twee relaties waarmee documenten aan elkaar gelinkt kunnen worden.

| Relatie | Beschrijving | Rol1 | Rol2 | MDTO-relatie |
| --- | ---| ---| ---| ---|
| heeft_publiceerbare_versie |  Heeft publiceerbare versie. | niet_publiceerbare_versie | publiceerbare_versie | heeft_versie |
| is_vervangen_door_gestempelde_versie | Is vervangen door gestempelde versie. | ongestempelde_versie | gestempelde_versie | is_vervangen_door |

De kolom Relatie geeft de relatie weer tussen twee informatieobjecten, zeg `informatieobject1` en `informatieobject2`. Kolommen Rol1 en Rol2 geven de rollen weer die respectievelijk `informatieobject1` en `informatieobject2` spelen in deze relatie.

## POST

In de volgende POST-operatie wordt de relatie vastgelegd dat document `<doc1>` is vervangen door de gestempelde versie `<doc2>`. De notatie `<doc>` staat voor de url link naar document `doc`.

`POST /relaties`

```json
{
    "informatieobject1": "<doc1>",
    "informatieobject2": "<doc2>",
    "relatie": "is_vervangen_door_gestempelde_versie"
}
```

Respons:

```json
{
    "url": "<rel1>",
    "uuid": "<<rel1>>",
    "informatieobject1": "<doc1>",
    "informatieobject2": "<doc2>",
    "relatie": "is_vervangen_door_gestempelde_versie",
} 
```
Aan de objecten in het responsbericht zijn zowel een url naar het object zelf en een uuid ter identificatie van object toegevoegd. De notatie `<<rel>>` staat voor de uuid van relatie `rel`.

In de volgende POST-opdracht wordt vastgelegd dat `<doc2>` een publiceerbare versie `<doc3>` heeft.

`POST /relaties`

```json
{
    "informatieobject1": "<doc2>",
    "informatieobject2": "<doc3>",
    "relatie": "heeft_publiceerbare_versie"
}
```

Respons:

```json
{
    "url": "<rel2>",
    "uuid": "<<rel2>>",
    "informatieobject1": "<doc2>",
    "informatieobject2": "<doc3>",
    "relatie": "heeft_publiceerbare_versie",
}
```

## GET

Alle attributen van de resource `/relaties` mogen gebruikt worden als query parameter voor filtering:

- `url`
- `uuid`
- `informatieobject1`
- `informatieobject2`
- `relatie`

Er is één extra query parameter toevoegd zodat het opgegeven infomatieobject zowel kan matchen op `ìnformatieobject1` als op `informatieobject2`:

- `informatieobject`

Bijvoorbeeld, onderstaande query geeft alle relaties terug waarin `doc2` voorkomt. 

`GET /relaties?informatieobject=<doc2>`

Respons:

```json
[
    {
        "url": "<rel1>",
        "uuid": "<<rel1>>",
        "informatieobject1": "<doc1>",
        "informatieobject2": "<doc2>",
        "relatie": "is_vervangen_door_gestempelde_versie",
    },
    {
        "url": "<rel2>",
        "uuid": "<<rel2>>",
        "informatieobject1": "<doc2>",
        "informatieobject2": "<doc3>",
        "relatie": "heeft_publiceerbare_versie",
    }
]
```

## DELETE

De volgende DELETE-operatie verwijderd relatie `rel2`.

`DELETE /relaties/<<rel2>>`

## PUT/PATCH

Er worden geen PUT/PATCH-operaties ondersteund. Updates kunnen alleen worden uitgevoerd door middel van DELETE en POST. Dus om een bestaande relatie tussen documenten aan te passen wordt eerst de relatie verwijderd met een DELETE en vervolgens wordt de aanpassing als een nieuwe relatie opgevoerd met een POST.

## Convenience

Voor het gemak is de `/enkelvoudiginformatieobjecten` resource uitgebreid met het attribuut `"gerelateerdeInformatieobjecten"`. Dit is een read-only veld dat alle gerelateerde informatieobjecten van een (enkelvoudig) informatieobject toont. Hieronder worden de drie eerder besproken documenten bevraagd door middel van een GET-operatie en kunnen we direct hun onderlinge relaties bekijken. Op basis van de referentietabel wordt de `rol` van het gerelateerde informatieobject in de relatie toegevoegd.

`GET /enkelvoudiginformatieobjecten/<<doc1>>`

Respons:

```json
[
    {
        "url": "<doc1>",
        "uuid": "<<doc1>>",
        ...
        "gerelateerdeInformatieobjecten":
            [
                {
                    "informatieobject": "<doc2>",
                    "rol": "gestempelde_versie",
                    "relatie": "<rel1>"
                },
            ]
    }
]
```

`GET /enkelvoudiginformatieobjecten/<<doc2>>`

Respons:

```json
[
    {
        "url": "<doc2>",
        "uuid": "<<doc2>>",
        ...
        "gerelateerdeInformatieobjecten":
            [
                {
                    "informatieobject": "<doc1>",
                    "rol": "ongestempelde_versie",
                    "relatie": "<rel1>"
                },
                {
                    "informatieobject": "<doc3>",
                    "rol": "publiceerbare_versie",
                    "relatie": "<rel2>"
                }
            ]
    }
]
```

`GET /enkelvoudiginformatieobjecten/<<doc3>>`

Respons:

```json
[
    {
        "url": "<doc3>",
        "uuid": "<<doc3>>",
        ...
        "gerelateerdeInformatieobjecten":
            [
                {
                    "informatieobject": "<doc2>",
                    "rol": "ongepubliceerde_versie",
                    "relatie": "<rel2>"
                }
            ]
    }
]
```

**Comment by MNIJ** [comment]
Dit lijkt me qua structuur een goed en werkbaar voorstel.

Het lijkt me wel verstandig om in de tabel met mogelijke waarden ook een generieke waarde op te nemen `is_gerelateerd_aan` met zowel voor rol1 als rol2 `gerelateerd_document`. Deze generieke waarde kan dan gebruikt worden voor use-cases waar (nog) geen specifiek relatietype is opgenomen in de referentietabel.
Verder zou ik ook `is_vervangen_door` opnemen. Deze waarde kan bijvoorbeeld gebruikt worden als een zaakbetrokkene (bijvoorbeeld een aanvragen van een vergunning) een nieuw document aanlevert ter vervanging van een bestaand document.

Van de voorgestelde waarde `is_vervangen_door_gestempelde_versie` vraag ik me af of deze in de praktijk gebruikt gaat worden. Door gesprekken met gemeenten en ook de opmerkingen van diverse gemeenten in #2079 zijn wij inmiddels tot de conclusie gekomen dat stempelen in principe niet tot een nieuw Enkelvoudiginformatieobject zou moeten leiden, maar "gewoon" een nieuwe versie van het bestaande Enkelvoudiginformatieobject. Dat is alleen vaak niet mogelijk vanwege de business rules in de DRC, vandaar ook ons wijzigingsvoorstel #2241.

**Comment by MNIJ** [comment]
Dit voorstel is kort besproken in een meeting op 17-10-2023. Uitkomst daarvan is o.a. dat we duidelijk onderscheid moeten maken in de definities van de concepten "versie" en "variant". Zoals afgesproken doe ik een voorstel voor deze definities binnen de DRC.
### Definities
**Versie**
Een exemplaar van een `informatieobject` ontstaan in de levensloop van dit  `informatieobject`. Elke wijziging aan een `informatieobject`, zowel in metadata als bestandsinhoud, levert een nieuwe versie op. Daarmee heeft elk  `informatieobject` een opeenvolgende reeks van versies. In het normale gebruik is alleen de meest recente versie relevant. Eerdere versies zijn alleen relevant in het kader van historie en audit logging.

**Variant**
Een alternatief van een `informatieobject`. Verschillende varianten kunnen gelijktijdig relevant zijn en elke variant is daarom een opzichzelfstaand `informatieobject`.

### Impact op dit voortel
Het concept `versie` zoals hierboven gedefinieerd is de bestaande werking van de DRC. Dit concept mogen we niet vermengen met het relateren van verschillende informatieobjecten, de term `versie` kan dus niet gebruikt worden in de beschrijving van de relatie tussen informatieobjecten. Concreet moet daarom de term `heeft_publiceerbare_versie` (en de afgeleiden daarvan) aangepast worden naar `heeft_publiceerbare_variant`.  In de publiceerbare variant kan immers data weggelakt zijn in het kader van anonimisering of bescherming van vertrouwelijke informatie, die wel degelijk nog relevant is voor de behandeling van het document. Dit valt daarmee onder de bovenstaande definitie van varianten.

Bij het stempelen van documenten is de consensus dat dit valt onder de definitie van versies en niet van varianten (zie ook de opmerking van 31 mei) en daarmee al gedekt is door de bestaande werking van de DRC, mede door wijziging #2241 die beschikbaar is gekomen in DRC 1.4.0. Daarom komt de behoefte aan het relatietype `is_vervangen_door_gestempelde_versie` te vervallen.

Het is daarmee niet gezegd dat aan elkaar gerelateerde informatieobjecten in alle gevallen varianten van elkaar zijn, er zijn ook andere relaties denkbaar (bijvoorbeeld "is_vervangen_door" of "is_beantwoord_met"), maar het zijn in elk geval nooit versies van elkaar in de terminologie van de DRC.

**Comment by hdksi** [comment]
Na een beperkte zoekactie kan ik niet vaststellen dat gebruik van het begrip 'variant' in relatie tot documenten gebruikelijk is. De noodzaak voor onderscheid tussen versies en varianten ontstaat echter vooral door de smalle definitie in het [RGBZ](https://www.gemmaonline.nl/index.php/Rgbz_2.0/doc/attribuutsoort/informatieobject.versie) ("aanduiding van de bewerkingsfase van het informatieobject").

Een alternatief is voor het introduceren van het begrip 'variant', is het uitgaan van de bredere definitie van 'versie' zoals gehanteerd binnen [Dublin Core](https://standaarden.overheid.nl/owms/4.0/doc/eigenschappen/dcterms.hasversion) en door overerving ook [MDTO](https://www.nationaalarchief.nl/archiveren/mdto/begrippenlijst-metagegevensschema#collapse-102683) ("informatieobject dat een versie of aanpassing is van een ander (informatie)object"). Dit betekent wel dat voor de attribuutsoort 'versie' in de Documenten API expliciet moet worden gemaakt dat die betrekking heeft op de 'bewerkingstijdslijn'. Hernoeming hiervan ligt daarbij voor de hand.

Toegevoegde rationale: voordeel van dit alternatief kan zijn dat we geen inspanning hoeven doen om een conceptueel juiste, maar 'synthetische' scheidslijn tussen twee concepten aan te brengen die de buitenwereld als één ziet: het zijn 'gewoon' allemaal versies, waarbij de relatie duidelijk maakt hoe het karakter van een specifieke versie ten opzichte van hetgeen waarvan die is afgeleid begrepen moet worden.

**Comment by michielverhoef** [comment]
Ondanks dat ik snap dat de term `versie`  gebruikelijk is voor wat hierboven beschreven staat als `variant` vind ik het lastige van het gebruik van één term voor verschillende betekenissen is dat het niet gelijk duidelijk is wat er precies bedoeld wordt.  

Zou `Verschijningsvorm` een betere term zijn?

**Reply by hdksi** [reply]
> één term voor verschillende betekenissen

Als je de bredere definitie kiest en de bestaande vergeet is er nog maar één betekenis van het woord versie en zijn we 'in sync' met andere standaarden ;)

> Zou Verschijningsvorm een betere term zijn?

Dat is een synoniem van representatie. De tekst die ik hier schrijf kan ik ook representeren in een Pdf- of Wordbestand, maar omdat het woord representatie in zich houdt dat de inhoud van de representatie ten opzichte van hetgeen wordt gerepresenteerd niet verandert, geldt dat niet voor bijvoorbeeld een versie van deze tekst waar ik alle leestekens en lidwoorden uit heb geschrapt. Een geanonimiseerde versie zou ik dientengevolge niet beschouwen als verschijningsvorm of representatie van het origineel, waarmee dat begrip volgens mij niet het (hele) voorgestelde gebruik van variant afdekt.

**Comment by ArjanKloosterboer** [comment]
> Een alternatief is voor het introduceren van het begrip 'variant', is het uitgaan van de bredere definitie van 'versie' zoals gehanteerd binnen [Dublin Core] en door overerving ook [MDTO] ("informatieobject dat een versie of aanpassing is van een ander (informatie)object").

Consequentie is dan, lijkt me, dat in het voorstel voor statussen #2242 er meerdere versies van hetzelfde informatieobject geldig kunnen zijn (overeenkomend met de 'oorspronkelijke' variant en de 'afgeleide' varianten). Doorzien we dit en kunnen we dit tacklen?

**Comment by HenriKorver** [comment]
Naar aanleiding van gesprekken die gevoerd zijn in Nieuwegein  en voortschrijdend inzicht heb ik het API-ontwerp voor het relateren van informatieobjecten aangepast, zie [redoc](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/VNG-Realisatie/gemma-zaken/Documenten-relateren-aan-andere-relevante-documenten/api-specificatie/drc/current_version/openapi.yaml#tag/informatieobjectenrelaties/operation/informatieobjectrelatie_create). 

De naamgeving in deze nieuwe versie is consistent gemaakt met de naamgeving van het RGBZ waarin zaken aan elkaar gerelateerd kunnen worden met de relatie "heeft relevante andere".

![afbeelding](https://github.com/VNG-Realisatie/gemma-zaken/assets/37145898/b5fa783f-e61f-42a3-a7f1-a5ee9b4578dc)

Op dezelfde manier modelleren we de relatie "heeft relevante andere" tussen informatieobjecten:

![afbeelding](https://github.com/VNG-Realisatie/gemma-zaken/assets/37145898/87aa8534-97a6-49d0-b13d-aa8ca720876d)
De relatieklasse `INFORMATIEOBJECTENRELATIE` is vertaald naar het REST-endpoint `informatieobjectenrelaties` (in meervoud). 

In het volgende voorbeeld leggen we een relatie tussen twee documenten waarbij `<doc2>`  een publiceerbare variant is van `<doc1>`.

`POST /informatieobjectenrelaties`

```json
{
  "informatieobject": "<doc1>",
  "relevantAnderInformatieobject": "<doc2>",
  "aardRelatie": "publiceerbare_variant"
}
```

Deze relatie wordt als conveniance ook zichtbaar gemaakt in de resource `/enkelvoudiginformatieobject`:

`GET /enkelvoudiginformatieobjecten/<<doc1>>`

```
{
  ...,
  "relevanteAndereInformatieobjecten": [
    {
      "informatieobject": "<doc2>",
      "aardRelatie": "publiceerbare_variant",
      "richtingRelatie": "standaard"
    }
  ],
}
```

`GET /enkelvoudiginformatieobjecten/<<doc2>>`

```
{
  ...,
  "relevanteAndereInformatieobjecten": [
    {
      "informatieobject": "<doc1>",
      "aardRelatie": "publiceerbare_variant",
      "richtingRelatie": "omgekeerd"
    }
  ],
}
```

Het attribuut "richtingRelatie" wordt gebruikt om de waarde van het attribuut "aardRelatie" in de goede richting te lezen.

In de aanvullende specificatie van de standaard leggen we vast dat  het attribuut `aardRelatie` aanvankelijk alleen de twee volgende waarden mag hebben. 

| Waarde attribuut `aardRelatie` | Beschrijving |
| --- | ---|
| `publiceerbare_variant` | Het relevante andere informatieobject is een publiceerbare variant. |
| `ongespecificeerd` | De relatie met het relevante andere informatieobject is niet gespecificeerd. |

In de toekomst kan deze tabel eventueel worden uitgebreid met nieuwe waarden als daar behoefte aan is.

**Comment by HenriKorver** [comment]
Naar aanleiding van gesprekken met VNG-collega's heb ik het voorstel weer wat bijgeslepen. Ten eerste is de naamgeving van de modellering aangepast:

![afbeelding](https://github.com/VNG-Realisatie/gemma-zaken/assets/37145898/c0dbec32-14db-4e64-a015-c25ec2adc18e)

De relatieklasse `INFORMATIEOBJECTENRELATIE` is vertaald naar het REST-endpoint `informatieobjectenrelaties` (in meervoud).

In de aanvullende specificatie van de standaard leggen we vast dat het attribuut "relatie" de volgende waarden mag hebben.

| | Waarde attribuut "relatie"| Beschrijving | Inverse | Read-only |
| --- | --- | --- | --- | --- |
| 1 | `is_een_geanonimiseerde_variant_van`      | Informatieobject A is een geanonimiseerde variant van informatieobject B. | 3 | Nee |
| 2 | `is_een_vervanger_van`                    | Informatieobject A is een vervanger van informatieobject B.               | 4 | Nee |
| 3 | `heeft_als_geanonimiseerde_variant`       | Informatieobject A heeft als geanonimiseerde variant informatieobject B.  | 1 |  Ja |
| 4 | `is_vervangen_door`                       | Informatieobject A is vervangen door informatieobject B.                   | 2 |  Ja |

In de toekomst kan deze tabel eventueel worden uitgebreid met nieuwe waarden als daar behoefte aan is.

In het volgende voorbeeld leggen we vast dat `<doc2>` een geanonimiseerde variant is van `<doc1>`.

`POST /informatieobjectenrelaties`

```text
{
  "informatieobjectA": "<doc2>",
  "informatieobjectB": "<doc1>",
  "relatie": "is_een_geanonimiseerde_variant_van"
}
```

De notatie `<doc>` staat voor de de URL-referentie naar het informatieobject `doc`.

Deze relatie wordt als conveniance ook zichtbaar gemaakt in de resource `/enkelvoudiginformatieobject`:

`GET /enkelvoudiginformatieobjecten/<<doc2>>`

```text
{
  ...,
  "gerelateerdeInformatieobjecten": [
    {
      "informatieobject": "<doc1>",
      "relatie": "is_een_geanonimiseerde_variant_van"
    }
  ],
}
```

De notatie `<<doc>>` staat voor de uuid van het informatieobject `doc`.

`GET /enkelvoudiginformatieobjecten/<<doc1>>`

```text
{
  ...,
  "gerelateerdeInformatieobjecten": [
    {
      "informatieobject": "<doc2>",
      "relatie": "heeft_als_geanonimiseerde_variant"
    }
  ],
}
```

**Comment by HenriKorver** [comment]
Samenvattend zijn dit de verschillen:

- In het informatiemodel is het objecttype INFORMATIEOBJECT hernoemd naar ENKELVOUDIGINFORMATIEOBJECT voor consistentie met het RGBZ.
- Attribuut "aardRelatie" in het informatiemodel is hernoemd naar "relatie".
- De relatie "heeft relevante andere" in het informatiemodel is hernoemd naar "heeft gerelateerd".
- We maken expliciet gebruik van inverse relaties in plaats van het attribuut "richtingRelatie" om de richting van de relatie aan te duiden.
- In de resource `/informatieobjectenrelaties` is het attribuut "informatieobject" hernoemd naar "informatieobjectA" en het attribuut "relevantAnderInformatieobject" hernoemd naar "informatieobjectB".
- De relatie `heeft_als_publiceerbare_variant` is hernoemd naar de (inverse) relaties `is_een_geanonimiseerde_variant_van` en  `heeft_als_geanonimiseerde_variant`.
- De (inverse) relaties  `is_een_vervanger_van` en `is_vervangen_door` zijn toegevoegd.
- De relatie `ongespecificeerd` is verwijderd om wildgroei te voorkomen.

**Comment by HenriKorver** [comment]
Inmiddels is het nieuwe voorstel uitgewerkt in de OAS, zie [redoc](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/VNG-Realisatie/gemma-zaken/Documenten-relateren-aan-andere-relevante-documenten/api-specificatie/drc/current_version/openapi.yaml#tag/informatieobjectenrelaties/operation/informatieobjectrelatie_create).

Graag jullie feedback of goedkeuring!  Als dat binnen is, zal ik ook het expand-mechanisme toevoegen.

**Comment by w-van-weelden** [comment]
De relatie 'is_een_geanonimiseerde_variant_van' vind ik niet helemaal gelukkig. Anonimiseren is maar één van de gronden waarop (delen van) documenten worden gelakt. Op deze manier zou je voor elke uitzonderingsgrond een aparte relatie moeten kunnen leggen. En een document kan meer dan één uitzonderingsgrond bevatten. Nu zijn persoonsgegevens in de praktijk de meest voorkomende grond, maar dat wil niet zeggen dat we daar het model op moeten baseren. De publiceerbare_variant relatie was beter. Maar ook dit past niet helemaal, omdat sommige documenten ongelakt gepubliceerd kunnen worden. Dan zou ik 'is_een_gelakte_variant_van' als relatienaam willen voorstellen.

De discussie of een informatieobject een enkelvoudiginformatieobject moet zijn, hoort hier ook bij. Want een document waarvan een gelakte variant bestaat, zou volgens mij in MDTO niet enkelvoudig zijn. Het informatieobject zelf bevat één of meer bestanden. Waarbij de gelakte versie een (hiërarchische) relatie heeft met het 'formele' informatieobject. Het hernoemen in het informatiemodel lijkt me onjuist. In RGBZ-termen zou ik eerder verwachten dat een document met een gelakte versie een samengesteld document is. Of zie ik dat verkeerd?

Tot slot zou het een mooie uitbreiding zijn om naast het vastleggen van gelakte documenten ook bij het informatieobject de uitzonderingsgronden voor publicatie te kunnen maken. Dan is ook in de keten de reden duidelijk dat een informatieobject niet (volledig) kan worden vrijgegeven.

**Comment by hdksi** [comment]
> De relatie 'is_een_geanonimiseerde_variant_van' vind ik niet helemaal gelukkig. Anonimiseren is maar één van de gronden waarop (delen van) documenten worden gelakt. Op deze manier zou je voor elke uitzonderingsgrond een aparte relatie moeten kunnen leggen. En een document kan meer dan één uitzonderingsgrond bevatten. Nu zijn persoonsgegevens in de praktijk de meest voorkomende grond, maar dat wil niet zeggen dat we daar het model op moeten baseren. De publiceerbare_variant relatie was beter. 
Maar ook dit past niet helemaal, omdat sommige documenten ongelakt gepubliceerd kunnen worden. Dan zou ik 'is_een_gelakte_variant_van' als relatienaam willen voorstellen.

Intern hebben we hierover ook discussie gehad. Inderdaad drukt de relatieaanduiding `is publiceerbare variant van` niet helder uit of er tussen de twee gerelateerde informatieobjecten een inhoudelijk verschil bestaat, en zo ja, waar nu inhoudelijk dat verschil uit bestaat. De naam `is geanonimiseerde variant van` is weliswaar duidelijker, maar zoals je terecht stelt geeft de afwezigheid van persoonsgegevens je niet automatisch de mogelijkheid de inhoud van een informatieobject openbaar te maken. En andersom betekent de aanwezigheid van persoonsgegevens ook niet zonder meer dat de inhoud van een informatieobject niet openbaar mag worden gemaakt.

Het alternatief `is gelakte variant van` vind ik een goede suggestie, maar eveneens lastig, omdat de relatie in dit geval bestaat bij de gratie van een technische maatregel ('het lakken') terwijl er ook andere mogelijkheden zijn om passages die niet openbaar gemaakt mogen worden niet openbaar te maken.

Als we de 'geanonimiseerde' optie omdat die niet aansluit bij het bedoelde gebruik even vergeten, hebben we nu kandidaten voor de relatienaam die een maatregel uitdrukken ('het lakken') en het resultaat van die maatregel ('het is publiceerbaar').

Wat we eigenlijk willen uitdrukken is de áfwezigheid van elementen in de inhoud van het openbaar te maken informatieobject die (volledige) openbaarmaking van het origineel in de weg staan. Probleem daarbij is dat er diverse grondslagen kunnen zijn voor die niet-(volledige)openbaarmaking. Bijvoorbeeld de in Wet open overheid genoemde uitzonderingsgronden, maar ook, voor zover die van invloed zijn op openbaarmaking, auteursrechtelijke of contractueel overeengekomen gebruiksbeperkingen. Volgens mij kunnen we deze samenvatten als 'beperkingen aan de openbaarheid'.

Als we bovendien duidelijk willen maken dat het bestaan van beperkingen in 'het origineel' een voorwaarde is voor het bestaan van een zonder beperkingen openbaar te maken variant, komen we uit bij iets als het volgende, helaas niet heel compacte, voorstel voor een relatiebeschrijving:

- [Informatieobject A] `is variant zonder openbaarheidsbeperkingen geldend voor` [informatieobject B], met als inverse 
- [informatieobject B] `heeft variant met opgeheven openbaarheidsbeperkingen in` [informatieobject A].

> De discussie of een informatieobject een enkelvoudiginformatieobject moet zijn, hoort hier ook bij. Want een document waarvan een gelakte variant bestaat, zou volgens mij in MDTO niet enkelvoudig zijn. Het informatieobject zelf bevat één of meer bestanden. Waarbij de gelakte versie een (hiërarchische) relatie heeft met het 'formele' informatieobject. Het hernoemen in het informatiemodel lijkt me onjuist. In RGBZ-termen zou ik eerder verwachten dat een document met een gelakte versie een samengesteld document is. Of zie ik dat verkeerd?

Dit vind ik een lastige vraag. MDTO stelt dat een 'bestand' zich verhoudt tot een 'informatieobject' doordat het (geparafraseerd) ["een representatie van een (deel van een) informatieobject is"](https://www.nationaalarchief.nl/archiveren/mdto/bestand). Persoonlijk kan ik me voorstellen dat varianten met en zonder openbaarheidsbeperkingen gezien worden als twee bestanden die één informatieobject representeren, maar ik weet niet of daarover consensus bestaat.

Het RGBZ kent het concept 'bestand' niet. Dit betekent dat we varianten met en zonder openbaarheidsbeperkingen sowieso als twee 'enkelvoudige informatieobjecten' moeten zien. Op basis van de [definitie voor samengesteld informatieobject](https://www.gemmaonline.nl/index.php/Rgbz_2.0/doc/objecttype/samengesteld_informatieobject) zou ik het niet gek vinden om de varianten middels dat concept samen te binden, ware het niet dat we in de Documenten API geen samengestelde informatieobjecten kennen.

> Tot slot zou het een mooie uitbreiding zijn om naast het vastleggen van gelakte documenten ook bij het informatieobject de uitzonderingsgronden voor publicatie te kunnen maken. Dan is ook in de keten de reden duidelijk dat een informatieobject niet (volledig) kan worden vrijgegeven.

Dit zou te realiseren zijn door de bedoeling en beschrijving van ['Gebruiksrechten informatieobject'](https://www.gemmaonline.nl/index.php/Rgbz_2.0/doc/gegevensgroeptype/informatieobject.gebruiksrechten_informatieobject), waarin voorwaarden voor raadplegen buiten beschouwing worden gelaten, uit te breiden richting ['beperkingGebruik' in MDTO](https://www.nationaalarchief.nl/archiveren/mdto/beperkingGebruik) waaronder wel expliciet openbaarheidsbeperkingen worden begrepen.

**Reply by w-van-weelden** [reply]
> [Informatieobject A] is variant zonder openbaarheidsbeperkingen geldend voor [informatieobject B], met als inverse
> [informatieobject B] heeft variant met opgeheven openbaarheidsbeperkingen in [informatieobject A].

De relatiebeschrijving lijkt me zo inderdaad veel beter recht te doen aan wat de relatie tussen twee documenten is. De naam is misschien wat cryptisch voor iemand die zicht het model eigen wil maken, maar past wel bij beter dan gelakte variant inderdaad. Dit lijkt mij een prima voorstel!

> Persoonlijk kan ik me voorstellen dat varianten met en zonder openbaarheidsbeperkingen gezien worden als twee bestanden die één informatieobject representeren, maar ik weet niet of daarover consensus bestaat.

Dit lijkt mij een zuivere weergave. Om het pragmatisch te houden, hoeven we dit ook niet hier op te lossen. Maar wellicht wel goed om dit mee te nemen bij een standaardmapping RGBZ <-> MDTO. 


> Persoonlijk kan ik me voorstellen dat varianten met en zonder openbaarheidsbeperkingen gezien worden als twee bestanden die één informatieobject representeren, maar ik weet niet of daarover consensus bestaat.

Dit lijkt me een elegente oplossing. Ik zou er wel voor willen pleiten om (in ieder geval voor de uitzonderingsgronden van de WOO) een enum te maken. Dat maakt rapportage (in de geest van de IHH-dashboards) op te publiceren documenten ook een stuk eenvoudiger.

**Comment by hdksi** [comment]
> Inmiddels is het nieuwe voorstel uitgewerkt in de OAS, zie [redoc](https://redocly.github.io/redoc/?url=https://raw.githubusercontent.com/VNG-Realisatie/gemma-zaken/Documenten-relateren-aan-andere-relevante-documenten/api-specificatie/drc/current_version/openapi.yaml#tag/informatieobjectenrelaties/operation/informatieobjectrelatie_create).

Mooi!

> De (inverse) relaties `is_een_vervanger_van` en `is_vervangen_door` zijn toegevoegd.

`is_een_vervanger_van` kan volgens mij vereenvoudigd worden: `vervangt`.
